### PR TITLE
feat(application): enable full 360 degree perception range

### PR DESCRIPTION
**Pull request by schwepmo**
## Type of this PR 

- [ ] Bug fix
- [x] Enhancement

## Description

* This PR enables the perception module to use full 360° surround perception
* This is done by extending the methodology applied in the first place by using the dot-product of vectors
* However, the dot-product only works for viewing angles less than 180°, so instead of comparing the left- and right-bound vectors, we do two checks:
   * 1st check if point is right of leftBound and left of headingVector
   * or 2nd check if point is left of rightBound and right of headingVector
* the image describes the process
360_degree_perception

* additionally we explicitly include points lying on the edges of the viewing range, which wasn't the case before due to floating point inaccuracies

## Affected parts of the online documentation

The perception module documentation should already be up to date

## Definition of Done

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All commits are signed-off (`-s`) with your mail address of your Eclipse Account.
- [x] `origin/main` has been merged into your Fork.
- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [x] There are no new SpotBugs warnings. 
- [x] Coding guidelines have been observed (see CONTRIBUTING.md).
- [x] All tests pass.

### CODE REVIEW

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/SimplePerceptionModule.java [hunk 1]
```diff
@@ -127,27 +133,49 @@ public boolean isInRange(SpatialObject other) {
...
                 // we use tmpVector2 as origin from the viewpoint of this object
                 tmpVector2.set(0, 0, 0);
-                return VectorUtils.isLeftOfLine(tmpVector1, tmpVector2, rightBoundVector) // vehicle is left of right edge
-                        && !VectorUtils.isLeftOfLine(tmpVector1, tmpVector2, leftBoundVector) // vehicle is right of left edge
-                        && tmpVector1.magnitude() <= configuration.getViewingRange(); // other vehicle is in range
+
+                if (tmpVector1.magnitude() > configuration.getViewingRange()) { // other vehicle is in range
+                    return false;
+                }
+                if (configuration.getViewingAngle() == 360) { // for 360 degree viewing angle field-of-view check is obsolete
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/SimplePerceptionModule.java]: Good point! This is fixed now.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/SimplePerceptionModule.java [hunk 2]
```diff
@@ -127,9 +133,53 @@ public boolean isInRange(SpatialObject other) {
...
+         * and checks if they point in the same direction (to avoid objects being located exactly in the opposite direction).
+         * If the magnitude is equal to 0 the point will be on the line, however could be anywhere in each direction, so
+         * we additionally check if the directions match.
+         * Note: This method only works, since we know that we work around the origin.
+         *
+         * @param pointToEvaluate point to be evaluated
+         * @param line            check point against this line
+         * @return {@code true} if point lies on line, otherwise {@code false}
+         */
+        private boolean liesOnVector(Vector3d pointToEvaluate, Vector3d line) {
```
Comment (schwepmo) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/SimplePerceptionModule.java:172]: This is a simplified version of [this](https://stackoverflow.com/a/11908158) answer.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/SimplePerceptionModule.java [hunk 3]
```diff
@@ -127,9 +133,53 @@ public boolean isInRange(SpatialObject other) {
                 other.getProjectedPosition().toVector3d(tmpVector1).subtract(origin);
                 // we use tmpVector2 as origin from the viewpoint of this object
                 tmpVector2.set(0, 0, 0);
-                return VectorUtils.isLeftOfLine(tmpVector1, tmpVector2, rightBoundVector) // vehicle is left of right edge
-                        && !VectorUtils.isLeftOfLine(tmpVector1, tmpVector2, leftBoundVector) // vehicle is right of left edge
-                        && tmpVector1.magnitude() <= configuration.getViewingRange(); // other vehicle is in range
+
+                if (tmpVector1.magnitude() > configuration.getViewingRange()) { // other vehicle is in range
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/SimplePerceptionModule.java:137]: sorry for being picky, but the comment is the negative of the condition 🤐

Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/SimplePerceptionModule.java]: Good Point! Fixed this

**Pull request merged**